src/track/playcounter.h

```diff
@@ -1,56 +1,107 @@
-#ifndef MIXXX_PLAYCOUNTER_H
-#define MIXXX_PLAYCOUNTER_H
+#pragma once
 
-#include "util/assert.h"
+#include <QDateTime>
+#include <QMetaType>
+#include <QtDebug>
 
+#include "util/assert.h"
 
-// Counts the total number of times a track has been played
-// and if the track has been played during the current session.
-class PlayCounter {
-public:
-    explicit PlayCounter(int timesPlayed = 0):
-        m_iTimesPlayed(timesPlayed),
-        m_bPlayed(false) {
+/// Counts the total number of times a track has been played
+/// and if the track has been played during the current session.
+class PlayCounter final {
+  public:
+    explicit PlayCounter(int timesPlayed = 0)
+            : m_timesPlayed(timesPlayed),
+              m_playedLatch(false) {
     }
 
-    // Sets total number of times a track has been played
-    void setTimesPlayed(int iTimesPlayed) {
-        DEBUG_ASSERT(0 <= iTimesPlayed);
-        m_iTimesPlayed = iTimesPlayed;
+    /// Set the total number of times a track has been played
+    /// while keeping the played flag and time stamp as is.
+    void setTimesPlayed(int timesPlayed) {
+        DEBUG_ASSERT(timesPlayed >= 0);
+        m_timesPlayed = timesPlayed;
     }
-    // Returns the total number of times a track has been played
+
+    /// Returns the total number of times a track has been played
     int getTimesPlayed() const {
-        return m_iTimesPlayed;
+        return m_timesPlayed;
+    }
+
+    /// Sets the (last) played at time stamp of a track without
+    /// affecting the play count.
+    void setLastPlayedAt(const QDateTime& lastPlayedAt) {
+        m_lastPlayedAt = lastPlayedAt;
     }
 
-    // Sets the played status of a track for the current session
-    // without affecting the play count.
-    void setPlayed(bool bPlayed = true) {
-        m_bPlayed = bPlayed;
+    /// Returns the actual time at which the track has been played
+    /// during this session.
+    ///
+    /// Note: Not stored persistently, so might be null or invalid
+    /// even if the corresponding boolean flags is true!
+    const QDateTime& getLastPlayedAt() const {
+        return m_lastPlayedAt;
     }
-    // Returns true if track has been played during the current session
+
+    /// Returns true if track is considered as played during the current session.
+    ///
+    /// This latch overrides the last played time stamp, which could never
+    /// be reset.
     bool isPlayed() const {
-        return m_bPlayed;
+        return m_playedLatch;
+    }
+
+    /// Sets or resets the played latch.
+    ///
+    /// The latch is also set implicitly to true when updating the
+    /// played at time stamp. It overrides the played at time stamp,
+    /// which could never be reset.
+    void setPlayedLatch(bool setPlayed) {
```

Comment:
A latch register is, unlike a normal flip-flop a memory that follows a value, but can be locked. I think we have no locking nature here, because the triggerLastPlayedNow is always possible. 

How about setPlayedFlag()?

Reply:
Flag is also ok.

For the future: Do we really need that magic? This custom flag will not available in aoide, only *times played* and *last played at*.

Reply:
I don't mind about the implementation, but I rely very much on this flag during a DJ session.